Refacto/threshold selection june21

slebastard opened the pull request:
This PR aims to:
- upgrade model evaluation method `evaluate` in `predictsignauxfaibles.evaluate`
- refactorize notebook `03-select-thresholds` into `03-select-thresholds-and-evaluate-model` that allow to simultaneously select ultimate hyperparameters (list levels "alert") and evaluate the model based on those thresholds

Simultaneously, a PR will be created to add:
- a documentation on threshold selection
- a documentation on model evaluation for the model of June 2021

slebastard committed:
refacto: updating notebook `03-select-thresholds` to use latest tools

Updating notebook `03-select-thresholds.ipynb` to use:
- function to merge multiple ModelPreds objects
- redressements (not yet integrated into our run() pipeline)

Splitting selection for models `default` and `small`

slebastard committed:
refacto: `utils.log_splits_size` now takes predictions numpy array as input

slebastard committed:
feat: Adding confusion matrix to the outputs of `evaluate.evaluate`

Using sklearn.metrics.confusion_matrix to output a dictionary `confusion_matrix` with keys `tn`, `fp`, `fn` and `tp`

slebastard committed:
refacto: notebook `03-select-thresholds`, adding model selector and list volumes comparison

slebastard committed:
feat: Adding AUCPR to the outputs of `evaluate.evaluate`

Adding Area Under Curve for Precision/Recall curve (AUCPR) to the outputs of `evaluate.evaluate`.
This metric can be effectively computed us the average precision score from sklearn.metric.average_precision_score

slebastard committed:
refacto: adding final documentation to 03-select-thresholds

slebastard committed:
refacto: renaming 03-select-thresholds --> 03-select-thresholds-and-evaluate-model

slebastard committed:
refacto: renaming arguments in `utils.log_splits_size`

slebastard committed:
refacto: `evaluate.make_thresholds_from_fbeta`

- Function `evaluate.make_thresholds_from_fbeta` now takes as inputs a pandas DataFrame, an array of true outcomes and a model pipeline as a `sklearn_pandas` Pipeline object. Method `predict_proba` from the model pipeline is used to compute predicted outcomes
- Disabling most logs from `sklearn_pandas` as they are significantly slowing down hyperparameter selection functions

slebastard committed:
feat: upgrading notebook 03-select-thresholds-and-evaluate-model with val/pred sets

Notebook 03-select-thresholds-and-evaluate-model now uses:
- a validation set `validate` to select classification thresholds
- a test set `test` to compute evaluaton metrics based on the selected hyperparameters
- a prediction set `predict` to measure the size of the future lists based on the selected hyperparameters

vviers commented on predictsignauxfaibles/evaluate.py:
```diff
@@ -128,24 +137,39 @@ def evaluate(
             If provided, the model will classify an entry X as positive if predict_proba(X)>=thresh.
             Otherwise, the model classifies X as positive if predict(X)=1, ie predict_proba(X)>=0.5
     """
+    aucpr = average_precision_score(
```
I don't think `average_precision_score` computes the AUCPR  (see https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html#sklearn.metrics.average_precision_score) so you might want to rename this variable to avoid confusion :)

If that's what you really want, AUCPR can probably be computed using `sklearn.metrics.auc` (which gives you the area under any curve)

slebastard replied:
Strictly speaking, it does compute the AUCPR, but in another way.
- **AUC with `sklearn.metrics.average_precision_score`**: Take the [formula of the average precision score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html#sklearn.metrics.average_precision_score):
$AP=\sum_{n}{(R_n - R_{n-1})*P_n}$
This is a Riemann integral computing the area under curve, without trapezoidal interpolation (ie, summing the areas under curve of each of the small rectangular blocks under the curve). This is an alternative way to compute the AUC.
- **AUC with `sklearn.metrics.auc`**: Function [`sklearn.metrics.auc`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.auc.html) also computes the area under curve with a Riemann integral, but using a trapezoïdal rule. This can be slightly optimistic, as described in the doc of `average_precision_score`.

However, since `sklearn.metrics.auc` is the most common implementation, and in order not to confuse users, I'm substituting `sklearn.metrics.auc` here

slebastard replied:
Actually let's stick to `average_precision_score` as its inputs are more intuitive, as they are the same as the other metrics:
- `average_precision_score` takes as inputs `y_true` and `y_score`, ie almost the same inputs as other metrics such as `balanced_accuracy_score`, `fbeta_score`, `precision_score`, etc
- `auc` takes as inputs a precision and a recall array, which means that both arrays would need to be recomputed within method `evaluate`, which would virtually be useless for the rest of the function

vviers replied:
Sounds good ! Thanks for clarifying, I learned something new today ! 🤓

vviers commented on predictsignauxfaibles/evaluate.py:
```diff
@@ -128,24 +137,39 @@ def evaluate(
     """
+    aucpr = average_precision_score(
+        dataset.data["outcome"],
+        model.predict_proba(dataset.data)[:, 1],
```
I would suggest that you compute `model.predict_proba(dataset.data)[:, 1]` once and for all at the beginning of your function, store it in a variable, and use the variable every time you need after that :)

This will save you (somewhat expensive) calls of the predict function

slebastard replied:
Yep, good idea!

slebastard replied:
Fixed in https://github.com/signaux-faibles/predictsignauxfaibles/pull/71/commits/4d423c9dc2736296841b82dba21a0160d6844f23

vviers reviewed (dismissed):
LGTM ! 💯

slebastard committed:
refacto: evaluate.py, computing y_true, y_score, y_pred ahead

At the beginning of method `predictsignauxfaibles.evaluate.evaluate`,
computing `y_true`, `y_score` (model score between 0 and 1) and `y_pred` (binary model predction) ahead,
as those three variables are used as such by most sklearn metrics.

vviers reviewed (commented):
LGTM ! 💯
